File: OBDII-Dashboard/src/main/java/com/sergiojosemp/obddashboard/service/ObdService.java
package com.sergiojosemp.obddashboard.service;

import android.app.IntentService;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Binder;
import android.os.IBinder;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.Nullable;

import com.sergiojosemp.obddashboard.R;
import com.sergiojosemp.obddashboard.activity.DashboardActivity;
import com.sergiojosemp.obddashboard.activity.DiagnosticTroubleCodeActivity;
import com.sergiojosemp.obddashboard.activity.Inject;
import com.sergiojosemp.obddashboard.activity.MenuActivity;
import com.sergiojosemp.obddashboard.activity.SettingsActivity;
import com.sergiojosemp.obddashboard.activity.VerboseActivity;
import com.github.pires.obd.commands.protocol.EchoOffCommand;
import com.github.pires.obd.commands.protocol.LineFeedOffCommand;
import com.github.pires.obd.commands.protocol.ObdResetCommand;
import com.github.pires.obd.commands.protocol.SelectProtocolCommand;
import com.github.pires.obd.commands.protocol.TimeoutCommand;
import com.github.pires.obd.commands.temperature.AmbientAirTemperatureCommand;
import com.github.pires.obd.enums.ObdProtocols;
import com.github.pires.obd.exceptions.MisunderstoodCommandException;
import com.github.pires.obd.exceptions.StoppedException;
import com.github.pires.obd.exceptions.UnsupportedCommandException;
import com.github.pires.obd.reader.ObdCommandJob;
import com.github.pires.obd.reader.ObdCommandJob.ObdCommandJobState;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

//Based on github.pires obd-reader  https://github.com/pires/android-obd-reader/

public class ObdService extends IntentService {

    private final String PREFERENCES = "preferences";
    private final IBinder binder = new ObdServiceBinder();
    // Vamos a definir una cola de trabajos como se hace en la aplicación sample de la API obd-java
    protected BlockingQueue<ObdCommandJob> jobsQueue = new LinkedBlockingQueue<>();
    private BlockingQueue<Boolean> obdDeviceQueue = new LinkedBlockingQueue<>();
    protected Long queueCounter = 0L;
    protected String TAG = ObdService.class.getName();
    protected Context ctx = this;
    private BluetoothDevice bluetoothDevice;
    private BluetoothSocket bluetoothSocket;
    private boolean verboseMode = false;
    private Boolean obdDevice = false;

    @Inject
    SharedPreferences preferences;
    private Thread t = new Thread(new Runnable() {
        @Override
        public void run() {
            try {
                Log.d(TAG, getText(R.string.obd_service_working).toString());
                executeQueue();
            } catch (InterruptedException e) {
                t.interrupt();
            }
        }
    });
    private Thread obdAdaptedCheckerThread = new Thread(new Runnable() {
        @Override
        public void run() {
            /*try {
                while (!obdAdaptedCheckerThread.isInterrupted()) {
                    if (obdDeviceQueue.take()) {//Blocks until someone puts a boolean value in the queue
                        ((MenuActivity) ctx).runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                ((MenuActivity) ctx).setObdIndicatorOn();
                            }
                        });
                        startObdConnection();
                    } else {
                        ((MenuActivity) ctx).runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                ((MenuActivity) ctx).setObdIndicatorOff();

                            }
                        });
                    }
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            catch (IOException e) {
                e.printStackTrace();
            }*/
        }
    });

    /**
     * Creates an IntentService.  Invoked by your subclass's constructor.
     *
     * @param name Used to name the worker thread, important only for debugging.
     */
    public ObdService(String name) {
        super(name);
    }
    public ObdService() {
        super("ObdService");
    }


    public SharedPreferences getPreferences() {
        return preferences;
    }

    public void setPreferences(SharedPreferences preferences) {
        this.preferences = preferences;
    }

    public void setVerboseMode(boolean verboseMode) {
        this.verboseMode = verboseMode;
    }

    public void putObdDeviceQueue(Boolean value) {
        try {
            obdDeviceQueue.put(value);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public void checkDevice() {
        if(!obdDevice && !obdAdaptedCheckerThread.isAlive()){ //Solo ejecutar el hilo si no se ha puesto en marcha anteriormente o si el que estaba en marcha ha terminado
            obdAdaptedCheckerThread.start();
        }
    }

    public synchronized void connectToDevice() {
        BluetoothSocket tmp = null;
        // Get a BluetoothSocket to connectToDevice with the given BluetoothDevice
        try {
            tmp = (BluetoothSocket) bluetoothDevice.getClass().getMethod("createRfcommSocket", new Class[]{int.class}).invoke(bluetoothDevice, 1);
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        }

        bluetoothSocket = tmp;

        try {
            bluetoothSocket.connect();
            Log.d(TAG, getText(R.string.connected_to_device).toString());
            //isObdDevice();
        } catch (IOException e) {
            e.printStackTrace();
            Toast.makeText(ctx, getText(R.string.connection_error_is_obd_adapter).toString(), Toast.LENGTH_SHORT).show();
        }
    }

    public void isObdDevice() throws IOException {

        Log.d(TAG, getText(R.string.checking_if_obdr).toString());
        if (bluetoothSocket.isConnected()) {
            final Thread obdCheck = new Thread(new Runnable() {
                @Override
                public void run() {
                    ObdCommandJob job = null;
                    job = new ObdCommandJob(new EchoOffCommand());
                    try {
                        Log.d(TAG, getText(R.string.dummy_command).toString());
                        job.getCommand().run(bluetoothSocket.getInputStream(), bluetoothSocket.getOutputStream()); //Blocking call
                        setObdDevice(true);
                        putObdDeviceQueue(true);
                        //startObdConnection();
                        Log.d(TAG, getText(R.string.is_obd).toString());
                    } catch (InterruptedException | IOException e) {
                        t.interrupt();
                    } catch(MisunderstoodCommandException me){
                        //Toast.makeText(ctx, "Error, please restart", Toast.LENGTH_SHORT).show();
                        t.interrupt();
                    } catch (StoppedException se){
                        t.interrupt();
                    }
                }
            });
            obdCheck.start();
        } else {
            Log.d(TAG, getText(R.string.status_bluetooth_error_connecting).toString());
            obdDevice = false;
            putObdDeviceQueue(false);
        }
    }

    private void startObdConnection() throws IOException {
        // Let's configure the connection.
        Log.d(TAG, getText(R.string.queueing_for_config).toString());
        jobsQueue.clear();
        queueJob(new ObdCommandJob(new ObdResetCommand()));
        queueJob(new ObdCommandJob(new ObdResetCommand()));
        queueJob(new ObdCommandJob(new ObdResetCommand()));

        //Below is to give the adapter enough time to reset before sending the commands, otherwise the first startup commands could be ignored.
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        queueJob(new ObdCommandJob(new EchoOffCommand()));
        queueJob(new ObdCommandJob(new LineFeedOffCommand()));
        queueJob(new ObdCommandJob(new TimeoutCommand(120))); //Según pruebas y documentación ELM327

        // Getting protocol from preferences
        final String protocol = preferences.getString(SettingsActivity.PROTOCOLS_LIST_KEY, "AUTO");
        queueJob(new ObdCommandJob(new SelectProtocolCommand(ObdProtocols.valueOf(protocol))));
        // Job for returning dummy data
        queueJob(new ObdCommandJob(new AmbientAirTemperatureCommand()));
        queueCounter = 0L;
        Log.d(TAG, getText(R.string.initialization_jobs).toString());
    }

    @Override
    public void onCreate() {
        super.onCreate();
        final BluetoothAdapter btAdapter = BluetoothAdapter.getDefaultAdapter();

        Log.d(TAG, getText(R.string.creating_service).toString());
        t.start();
        Log.d(TAG, getText(R.string.service_created).toString());
        preferences = getSharedPreferences(PREFERENCES,
                Context.MODE_MULTI_PROCESS);
    }

    public void queueJob(ObdCommandJob job) {
        queueCounter++;
        Log.d(TAG, getText(R.string.adding_job).toString() + queueCounter + getText(R.string.to_queue).toString());
        job.setId(queueCounter);
        try {
            jobsQueue.put(job);
            Log.d(TAG, getText(R.string.job).toString() + job.getCommand().getName() + getText(R.string.queued_succesfully).toString());
        } catch (InterruptedException e) {
            job.setState(ObdCommandJob.ObdCommandJobState.QUEUE_ERROR);
            Log.e(TAG, getText(R.string.failed_queueing_job).toString() + job.getCommand().getName());
        }
    }


    protected void executeQueue() throws InterruptedException {
        preferences = getSharedPreferences(PREFERENCES,
                Context.MODE_MULTI_PROCESS);

        while (!Thread.currentThread().isInterrupted()) {
            Log.d(TAG, getText(R.string.executing_queue).toString());
            ObdCommandJob job = null; // Un Job tiene un ObdCommand (comando cualquiera) y un estado (NEW, RUNNING,...)
            try {
                job = jobsQueue.take(); //Bloqueo hasta que empiecen a entrar comandos
                Log.d(TAG, getText(R.string.taking_job).toString() + job.getId() + getText(R.string.from_queue).toString());

                if (job.getState().equals(ObdCommandJobState.NEW) && (ctx.getClass().equals(DashboardActivity.class) || ctx.getClass().equals(VerboseActivity.class))) {
                    if (bluetoothSocket.isConnected() && (ctx.getClass().equals(DashboardActivity.class) || ctx.getClass().equals(VerboseActivity.class))) {
                        Log.d(TAG, getText(R.string.job_is_new).toString());
                        job.setState(ObdCommandJobState.RUNNING);
                        job.getCommand().run(bluetoothSocket.getInputStream(), bluetoothSocket.getOutputStream());
                        if (job != null) {
                            Log.d(TAG, getText(R.string.updating_dash).toString() + job.getCommand().getName() + getText(R.string.updating_dash).toString() + job.getCommand().getFormattedResult() + "...");
                            final ObdCommandJob job2 = job;
                            if(ctx != this) {/*
                                if (verboseMode) {
                                    ((VerboseActivity) ctx).runOnUiThread(new Runnable() {
                                        @Override
                                        public void run() {
                                            ((VerboseActivity) ctx).stateUpdate(job2);
                                        }
                                    });
                                } else {
                                    ((DashboardActivity) ctx).runOnUiThread(new Runnable() {
                                        @Override
                                        public void run() {
                                            ((DashboardActivity) ctx).stateUpdate(job2);
                                        }
                                    });
                                }*/
                            }
                        }
                    } else {
                        job.setState(ObdCommandJobState.EXECUTION_ERROR);
                        Log.e(TAG, getText(R.string.cant_run_closed_socket).toString());
                        Thread.currentThread().interrupt();
                    }
                } else {
                    Log.e(TAG, getText(R.string.bug_alert).toString());
                }
            } catch (InterruptedException i) {
                Thread.currentThread().interrupt();
            } catch (UnsupportedCommandException u) {
                if (job != null) {
                    job.setState(ObdCommandJobState.NOT_SUPPORTED);
                }
                Log.d(TAG, getText(R.string.command_not_supported).toString() + " -> " + u.getMessage());
            } catch (IOException io) {
                if (job != null) {
                    if (io.getMessage().contains("Broken pipe"))
                        job.setState(ObdCommandJobState.BROKEN_PIPE);
                    else
                        job.setState(ObdCommandJobState.EXECUTION_ERROR);
                }
                Log.e(TAG, getText(R.string.io_error).toString() + " -> " + io.getMessage());
            }catch (Exception e) {
                if (job != null) {
                    job.setState(ObdCommandJobState.EXECUTION_ERROR);
                }
                Log.e(TAG, getText(R.string.failed_to_run_command).toString() + " -> " + e.getMessage());
            }
        }
    }

    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }


    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        if (preferences == null) {
            preferences = getSharedPreferences(PREFERENCES,
                    Context.MODE_MULTI_PROCESS);
        }
    }

    @Override
    public void onDestroy() {
        Log.d(TAG, getText(R.string.service_done).toString());
    }

    public void setObdDevice(Boolean obdDevice) {
        this.obdDevice = obdDevice;
    }

    public void setBluetoothDevice(BluetoothDevice btd) {
        this.bluetoothDevice = btd;
    }

    public boolean queueEmpty() {
        return jobsQueue.isEmpty();
    }

    public void emptyQueue() {
        jobsQueue.clear();
    }

    public BluetoothSocket getbluetoothSocket() {
        return bluetoothSocket;
    }

    public void setContext(Context c) {
        ctx = c;
    }

    public void startService() {
        if (preferences == null) {
            preferences = getSharedPreferences(PREFERENCES,
                    Context.MODE_MULTI_PROCESS);
        }
        Log.d(TAG, getText(R.string.starting_obd_service).toString());
        if(ctx != null && ctx.getClass().equals(DiagnosticTroubleCodeActivity.class)) {
            Log.d(TAG,getText(R.string.executing_queue_paused).toString());
            if(bluetoothSocket != null && bluetoothSocket.isConnected()) {
                troubleCodes = requestTroubleCodes();
            }else{
                Toast.makeText(ctx, getText(R.string.error_getting_dtc).toString(), Toast.LENGTH_LONG).show();
            }
        }

    }

    public String getTroubleCodes() {
        return troubleCodes;
    }

    public void setTroubleCodes(String troubleCodes) {
        this.troubleCodes = troubleCodes;
    }

    private String troubleCodes = "";

    public void stopService() {
    }

    //Devuelve una instancia de esta misma clase, así quien se enlace con este servicio, puede acceder a los métodos públicos
    // de esta clase
    public class ObdServiceBinder extends Binder {
        public ObdService getService() {
            return ObdService.this;
        }
    }


    private boolean dtc = false;

    public boolean isDtc() {
        return dtc;
    }

    public void setDtc(boolean dtc) {
        this.dtc = dtc;
    }

    public String requestTroubleCodes(String... params) {
        String result = "";
        synchronized (this) {
            try {
                // Let's configure the connection.
                Log.d(TAG, getText(R.string.queueing_for_config).toString());
                new ObdResetCommand().run(bluetoothSocket.getInputStream(), bluetoothSocket.getOutputStream());
                new EchoOffCommand().run(bluetoothSocket.getInputStream(), bluetoothSocket.getOutputStream());
                new LineFeedOffCommand().run(bluetoothSocket.getInputStream(), bluetoothSocket.getOutputStream());
                new SelectProtocolCommand(ObdProtocols.AUTO).run(bluetoothSocket.getInputStream(), bluetoothSocket.getOutputStream());


                DiagnosticTroubleCodeActivity.ModifiedTroubleCodesObdCommand tcoc = new DiagnosticTroubleCodeActivity.ModifiedTroubleCodesObdCommand();
                tcoc.run(bluetoothSocket.getInputStream(), bluetoothSocket.getOutputStream());
                result = tcoc.getFormattedResult();


            } catch (IOException e) {
                e.printStackTrace();
                Log.e(TAG, e.getMessage());
                return null;
            } catch (InterruptedException e) {
                e.printStackTrace();
                Log.e(TAG, e.getMessage());
                return null;
            }
        }
        return result;
    }

    public Context getDefaultContext(){
        return this;
    }
}

File: OBDII-Dashboard/src/main/java/com/sergiojosemp/obddashboard/service/ObdService.kt
package com.sergiojosemp.obddashboard.service

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothSocket
import android.content.Context
import android.os.Binder
import android.util.Log
import android.widget.Toast
import androidx.lifecycle.LifecycleService
import com.github.pires.obd.commands.protocol.EchoOffCommand
import com.github.pires.obd.commands.protocol.LineFeedOffCommand
import com.github.pires.obd.commands.protocol.ObdResetCommand
import com.github.pires.obd.commands.protocol.SelectProtocolCommand
import com.github.pires.obd.commands.temperature.AmbientAirTemperatureCommand
import com.github.pires.obd.commands.control.TroubleCodesCommand
import com.github.pires.obd.enums.ObdProtocols
import com.github.pires.obd.exceptions.MisunderstoodCommandException
import com.github.pires.obd.exceptions.StoppedException
import com.github.pires.obd.exceptions.UnsupportedCommandException
import com.github.pires.obd.reader.ObdCommandJob
import com.sergiojosemp.obddashboard.R
import com.sergiojosemp.obddashboard.activity.DiagnosticTroubleCodeActivity
import com.sergiojosemp.obddashboard.model.PreferencesHelper
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.io.IOException

class ObdService : LifecycleService() {

    private val TAG = ObdService::class.java.name

    private val serviceScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val binder = ObdServiceBinder()
    private var commandChannel: Channel<ObdCommandJob> = Channel()

    private var jobsQueueCounter = 0L
    private var bluetoothDevice: BluetoothDevice? = null
    private var bluetoothSocket: BluetoothSocket? = null
    private var verboseMode = false
    private var obdDevice = false
    private var dtc = false
    private var troubleCodes = ""

    @Volatile
    private var contextRef: Context? = null

    // ModifiedTroubleCodesObdCommand nested class for DTC retrieval
    inner class ModifiedTroubleCodesObdCommand : TroubleCodesCommand() {
        override fun getResult(): String {
            return rawData.replace("SEARCHING...", "").replace("NODATA", "")
        }
    }

    override fun onCreate() {
        super.onCreate()
        Log.d(TAG, getText(R.string.creating_service).toString())
        serviceScope.launch {
            executeQueue()
        }
        Log.d(TAG, getText(R.string.service_created).toString())
    }

    override fun onDestroy() {
        super.onDestroy()
        commandChannel.close()
        serviceScope.cancel()
        Log.d(TAG, getText(R.string.service_done).toString())
    }

    /**
     * Queues an OBD command job for execution. Assigns a unique ID and adds it to the channel.
     */
    fun queueJob(job: ObdCommandJob) {
        synchronized(this) {
            jobsQueueCounter++
        }.also { id ->
            Log.d(
                TAG,
                getText(R.string.adding_job).toString() + id + getText(R.string.to_queue).toString()
            )
            job.id = id
            try {
                runBlocking { commandChannel.send(job) }
                Log.d(
                    TAG,
                    getText(R.string.job).toString() + job.command.getName() + getText(R.string.queued_succesfully).toString()
                )
            } catch (e: InterruptedException) {
                job.setState(ObdCommandJob.ObdCommandJobState.QUEUE_ERROR)
                Log.e(TAG, getText(R.string.failed_queueing_job).toString() + job.command.getName())
            }
        }
    }

    /**
     * Core command execution loop. Reads jobs from the channel and executes them on the BT socket.
     */
    private suspend fun executeQueue() {
        Log.d(TAG, getText(R.string.obd_service_working).toString())

        for (job in commandChannel) {
            try {
                Log.d(
                    TAG,
                    getText(R.string.taking_job).toString() + job.id + getText(R.string.from_queue).toString()
                )

                if (job.state == ObdCommandJob.ObdCommandJobState.NEW &&
                    bluetoothSocket?.isConnected == true) {
                    Log.d(TAG, getText(R.string.job_is_new).toString())
                    job.setState(ObdCommandJob.ObdCommandJobState.RUNNING)
                    job.command.run(bluetoothSocket!!.getInputStream(), bluetoothSocket!!.getOutputStream())
                    if (job != null) {
                        Log.d(
                            TAG,
                            getText(R.string.updating_dash).toString() +
                                    job.command.getName() +
                                    getText(R.string.updating_dash).toString() +
                                    job.command.getFormattedResult() + "..."
                        )
                    }
                } else {
                    if (bluetoothSocket?.isConnected != true) {
                        job.setState(ObdCommandJob.ObdCommandJobState.EXECUTION_ERROR)
                        Log.e(TAG, getText(R.string.cant_run_closed_socket).toString())
                    } else {
                        Log.e(TAG, getText(R.string.bug_alert).toString())
                    }
                }
            } catch (e: UnsupportedCommandException) {
                job?.setState(ObdCommandJob.ObdCommandJobState.NOT_SUPPORTED)
                Log.d(TAG, getText(R.string.command_not_supported).toString() + " -> " + e.message)
            } catch (e: IOException) {
                if (job != null) {
                    if (e.message?.contains("Broken pipe") == true) {
                        job.setState(ObdCommandJob.ObdCommandJobState.BROKEN_PIPE)
                    } else {
                        job.setState(ObdCommandJob.ObdCommandJobState.EXECUTION_ERROR)
                    }
                }
                Log.e(TAG, getText(R.string.io_error).toString() + " -> " + e.message)
            } catch (e: Exception) {
                job?.setState(ObdCommandJob.ObdCommandJobState.EXECUTION_ERROR)
                Log.e(TAG, getText(R.string.failed_to_run_command).toString() + " -> " + e.message)
            }
        }
    }

    /**
     * Establishes a Bluetooth RFCOMM connection to the previously set device on channel 1.
     */
    @Synchronized
    fun connectToDevice() {
        var tmp: BluetoothSocket? = null
        try {
            tmp = bluetoothDevice?.javaClass?.getMethod("createRfcommSocket", Int::class.javaPrimitiveType)
                ?.invoke(bluetoothDevice, 1) as? BluetoothSocket
        } catch (e: Exception) {
            e.printStackTrace()
        }

        bluetoothSocket = tmp

        try {
            bluetoothSocket?.connect()
            Log.d(TAG, getText(R.string.connected_to_device).toString())
        } catch (e: IOException) {
            e.printStackTrace()
            Toast.makeText(contextRef, getText(R.string.connection_error_is_obd_adapter), Toast.LENGTH_SHORT).show()
        }
    }

    /**
     * Sends a dummy EchoOffCommand to verify the connected device is an OBD adapter.
     */
    @Throws(IOException::class)
    fun isObdDevice() {
        Log.d(TAG, getText(R.string.checking_if_obdr).toString())
        if (bluetoothSocket?.isConnected == true) {
            serviceScope.launch {
                try {
                    Log.d(TAG, getText(R.string.dummy_command).toString())
                    val job = ObdCommandJob(EchoOffCommand())
                    runBlocking {
                        job.command.run(
                            bluetoothSocket!!.getInputStream(),
                            bluetoothSocket!!.getOutputStream()
                        )
                    }
                    setObdDevice(true)
                    putObdDeviceQueue(true)
                    Log.d(TAG, getText(R.string.is_obd).toString())
                } catch (e: MisunderstoodCommandException) {
                    e.printStackTrace()
                } catch (e: StoppedException) {
                    e.printStackTrace()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        } else {
            Log.d(TAG, getText(R.string.status_bluetooth_error_connecting).toString())
            obdDevice = false
            putObdDeviceQueue(false)
        }
    }

    /**
     * Queues initialization commands (reset, echo off, line feed off, timeout, protocol select).
     */
    @Throws(IOException::class)
    private fun startObdConnection() {
        Log.d(TAG, getText(R.string.queueing_for_config).toString())
        emptyQueue()
        queueJob(ObdCommandJob(ObdResetCommand()))
        queueJob(ObdCommandJob(ObdResetCommand()))
        queueJob(ObdCommandJob(ObdResetCommand()))

        try {
            Thread.sleep(1000)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

        queueJob(ObdCommandJob(EchoOffCommand()))
        queueJob(ObdCommandJob(LineFeedOffCommand()))
        queueJob(ObdCommandJob(com.github.pires.obd.commands.protocol.TimeoutCommand(120)))

        val protocol = PreferencesHelper.getStringSync(this, PreferencesHelper.PROTOCOLS_LIST_KEY.toString(), "AUTO")
        queueJob(ObdCommandJob(SelectProtocolCommand(ObdProtocols.valueOf(protocol))))

        queueJob(ObdCommandJob(AmbientAirTemperatureCommand()))
        jobsQueueCounter = 0L
        Log.d(TAG, getText(R.string.initialization_jobs).toString())
    }

    /**
     * Called by Activities after binding. Triggers DTC retrieval if the bound context is DiagnosticTroubleCodeActivity.
     */
    fun startService() {
        Log.d(TAG, getText(R.string.starting_obd_service).toString())

        val ctx = contextRef
        if (ctx is DiagnosticTroubleCodeActivity && bluetoothSocket?.isConnected == true) {
            Log.d(TAG, getText(R.string.executing_queue_paused).toString())
            serviceScope.launch {
                troubleCodes = requestTroubleCodes() ?: ""
            }
        } else if (ctx is DiagnosticTroubleCodeActivity) {
            Toast.makeText(ctx, getText(R.string.error_getting_dtc), Toast.LENGTH_LONG).show()
        }
    }

    /**
     * Retrieves diagnostic trouble codes by sending configuration commands followed by a DTC query.
     * Returns null on I/O or interrupt errors. Uses runBlocking for BT operations to maintain API compatibility.
     */
    fun requestTroubleCodes(vararg params: String): String? {
        var result = ""
        synchronized(this) {
            try {
                Log.d(TAG, getText(R.string.queueing_for_config).toString())
                val input = bluetoothSocket?.getInputStream() ?: return null
                val output = bluetoothSocket?.getOutputStream() ?: return null

                runBlocking {
                    ObdResetCommand().run(input, output)
                    EchoOffCommand().run(input, output)
                    LineFeedOffCommand().run(input, output)
                    SelectProtocolCommand(ObdProtocols.AUTO).run(input, output)
                }

                val tcoc = ModifiedTroubleCodesObdCommand()
                runBlocking {
                    tcoc.run(input, output)
                }
                result = tcoc.getFormattedResult() ?: ""
            } catch (e: IOException) {
                e.printStackTrace()
                Log.e(TAG, e.message.toString())
                return null
            } catch (e: InterruptedException) {
                e.printStackTrace()
                Log.e(TAG, e.message.toString())
                return null
            }
        }
        return result
    }

    /**
     * Returns the currently stored trouble codes string.
     */
    fun getTroubleCodes(): String = troubleCodes

    /**
     * Sets the stored trouble codes string.
     */
    fun setTroubleCodes(troubleCodes: String) {
        this.troubleCodes = troubleCodes
    }

    fun stopService() {}

    // -- Binder for Activity binding --

    override fun onBind(intent: android.content.Intent): android.os.IBinder = binder

    /**
     * Binder returned to Activities. Provides access to the ObdService instance.
     */
    inner class ObdServiceBinder : Binder() {
        fun getService(): ObdService = this@ObdService
    }

    // -- Property getters/setters (API compatibility with Java callers) --

    fun setVerboseMode(verboseMode: Boolean) {
        this.verboseMode = verboseMode
    }

    fun putObdDeviceQueue(value: Boolean) {
        // Legacy method kept for API compatibility; no longer used with Channel-based queue.
    }

    /** @deprecated ObD adapted checker thread is no longer used in coroutine-based service. */
    @Deprecated("Use coroutine-based approach instead")
    fun checkDevice() {}

    fun setObdDevice(obdDevice: Boolean) {
        this.obdDevice = obdDevice
    }

    /**
     * Sets the Bluetooth device to connect to. Call before [connectToDevice].
     */
    fun setBluetoothDevice(btd: BluetoothDevice?) {
        this.bluetoothDevice = btd
    }

    /**
     * Returns true if there are no pending jobs in the command channel.
     */
    fun queueEmpty(): Boolean = commandChannel.isEmpty

    /**
     * Clears all pending jobs from the command channel by closing and recreating it.
     */
    fun emptyQueue() {
        val oldChannel = commandChannel
        oldChannel.close()
        commandChannel = Channel()
        serviceScope.launch { executeQueue() }
    }

    /**
     * Returns the currently connected Bluetooth socket, or null if not connected.
     */
    fun getbluetoothSocket(): BluetoothSocket? = bluetoothSocket

    /**
     * Sets the context reference from the binding Activity for UI-related operations.
     */
    fun setContext(c: Context?) {
        this.contextRef = c
    }

    fun isDtc(): Boolean = dtc

    fun setDtc(dtc: Boolean) {
        this.dtc = dtc
    }

    /**
     * Returns the service's own context (the Service itself).
     */
    fun getDefaultContext(): Context = this
}
